File: src/main/kotlin/lexer/Lexer.kt
package org.pawscript.lexer

import org.pawscript.error.PawError.LexError


class Lexer(private val source: String) {
    private val tokens = mutableListOf<Token>()
    private var start = 0
    private var current = 0
    private var line = 1
    private var col = 1
    private var inTemplate = false

    private val keywords = mapOf(
        "say" to TokenKind.SAY,
        "ask" to TokenKind.ASK,
        "let" to TokenKind.LET,
        "fun" to TokenKind.FUN,
        "if" to TokenKind.IF,
        "else" to TokenKind.ELSE,
        "loop" to TokenKind.LOOP,
        "in" to TokenKind.IN,
        "return" to TokenKind.RETURN,
        "bark" to TokenKind.BARK,
        "async" to TokenKind.ASYNC,
        "await" to TokenKind.AWAIT,
        "import" to TokenKind.IMPORT,
        "as"     to TokenKind.AS,
        "record" to TokenKind.RECORD,
        "tail" to TokenKind.TAIL,
        "sniff" to TokenKind.SNIFF,
        "snatch" to TokenKind.SNATCH,
        "lastly" to TokenKind.LASTLY,
        "break"    to TokenKind.BREAK,
        "continue" to TokenKind.CONTINUE,
        "nopaw" to TokenKind.NOPAW
    )

    /**
     * Tokenize the source into a list of tokens or throw LexError on failure.
     */
    @Throws(LexError::class)
    fun tokenize(): List<Token> {
        while (!isAtEnd()) {
            start = current
            scanToken()
        }
        tokens.add(Token(TokenKind.EOF, "", line, col))
        return tokens
    }

    private fun scanToken() {
        val c = advance()
        when {
            c == '\n' -> handleNewLine()
            c.isWhitespace() -> return
            c == '#' -> skipComment()
            inTemplate && c == '}' -> addToken(TokenKind.TEMPLATE_EXPR_END, "}")
            c == '`' -> templateStart()
            c == '"' -> stringLiteral()
            c == '\'' -> charLiteral()
            c.isLetter() || c == '_' -> identifierOrKeyword()
            c.isDigit() -> number()
            else -> symbolOrError(c)
        }
    }

    private fun handleNewLine() {
        line++
        col = 1
    }

    private fun skipComment() {
        while (peek() != '\n' && !isAtEnd()) advance()
    }

    private fun identifierOrKeyword() {
        while (peek().isLetterOrDigit() || peek() == '_') advance()
        emitKeywordOrIdentifier()
    }

    private fun emitKeywordOrIdentifier() {
        val text = source.substring(start, current)
        val kind = keywords[text] ?: TokenKind.IDENTIFIER
        addToken(kind, text)
    }

    private fun number() {
        while (peek().isDigit()) advance()
        if (peek() == '.' && peekNext().isDigit()) {
            advance()
            while (peek().isDigit()) advance()
            addToken(TokenKind.FLOAT_LITERAL, source.substring(start, current))
        } else {
            addToken(TokenKind.INT_LITERAL, source.substring(start, current))
        }
    }

    private fun stringLiteral() {
        val sb = StringBuilder()
        while (!isAtEnd() && peek() != '"') {
            if (peek() == '\\') {
                advance()
                sb.append(parseEscape())
            } else {
                sb.append(advance())
            }
        }
        if (isAtEnd()) {
            throw LexError("Unterminated string literal", line, col)
        }
        advance() // closing '"'
        addToken(TokenKind.STRING_LITERAL, sb.toString())
    }

    private fun charLiteral() {
        if (isAtEnd()) throw LexError("Unterminated char literal", line, col)
        val ch = if (peek() == '\\') {
            advance(); parseEscape()
        } else advance()
        if (peek() != '\'') throw LexError("Unterminated char literal", line, col)
        advance() // closing '\''
        addToken(TokenKind.CHAR_LITERAL, ch.toString())
    }

    private fun parseEscape(): Char {
        if (isAtEnd()) throw LexError("Invalid escape sequence at end of input", line, col)
        val escape = advance()
        return when (escape) {
            'n'  -> '\n'
            'r'  -> '\r'
            't'  -> '\t'
            '\\' -> '\\'
            '"' -> '"'
            '\''-> '\''
            '`'  -> '`'
            '$'  -> '$'
            else -> throw LexError("Invalid escape sequence: \\$escape", line, col)
        }
    }

    private fun templateStart() {
        inTemplate = true
        val sb = StringBuilder()
        while (!isAtEnd()) {
            when {
                peek() == '`' -> {
                    advance()
                    addToken(TokenKind.TEMPLATE_TEXT, sb.toString())
                    inTemplate = false
                    return
                }
                peek() == '$' && peekNext() == '{' -> {
                    addToken(TokenKind.TEMPLATE_TEXT, sb.toString())
                    sb.clear()
                    advance(); advance()
                    addToken(TokenKind.TEMPLATE_EXPR_START, "\${")
                    inTemplate = true
                    return
                }
                peek() == '\\' -> {
                    advance(); sb.append(parseEscape())
                }
                else -> sb.append(advance())
            }
        }
        throw LexError("Unterminated template literal", line, col)
    }

    private fun symbolOrError(c: Char) {
        when (c) {
            '+' -> addToken(TokenKind.PLUS)
            '-' -> if (match('>')) addToken(TokenKind.ARROW) else addToken(TokenKind.MINUS)
            '*' -> addToken(TokenKind.STAR)
            '/' -> addToken(TokenKind.SLASH)
            '%' -> addToken(TokenKind.PERCENT)
            '?' -> addToken(TokenKind.QUESTION)
            '(' -> addToken(TokenKind.LPAREN)
            ')' -> addToken(TokenKind.RPAREN)
            '{' -> addToken(TokenKind.LBRACE)
            '}' -> addToken(TokenKind.RBRACE)
            '[' -> addToken(TokenKind.LBRACKET)
            ']' -> addToken(TokenKind.RBRACKET)
            ',' -> addToken(TokenKind.COMMA)
            '.' -> if (match('.')) addToken(TokenKind.RANGE) else addToken(TokenKind.DOT)
            ':' -> addToken(TokenKind.COLON)
            ';' -> addToken(TokenKind.SEMICOLON)
            '!' -> if (match('=')) addToken(TokenKind.NEQ) else addToken(TokenKind.NOT)
            '=' -> if (match('=')) addToken(TokenKind.EQEQ) else addToken(TokenKind.ASSIGN)
            '<' -> if (match('=')) addToken(TokenKind.LE) else addToken(TokenKind.LT)
            '>' -> if (match('=')) addToken(TokenKind.GE) else addToken(TokenKind.GT)
            '&' -> if (match('&')) addToken(TokenKind.ANDAND) else throw LexError("Unexpected character '&'", line, col)
            '|' -> if (match('|')) addToken(TokenKind.OROR) else throw LexError("Unexpected character '|'", line, col)
            else -> throw LexError("Unexpected character '$c'", line, col)
        }
    }

    private fun addToken(kind: TokenKind, lexeme: String = source.substring(start, current)) {
        tokens.add(Token(kind, lexeme, line, col))
    }

    private fun match(expected: Char): Boolean {
        if (isAtEnd() || source[current] != expected) return false
        advance(); return true
    }

    private fun peek(): Char = if (isAtEnd()) '\u0000' else source[current]
    private fun peekNext(): Char = if (current + 1 >= source.length) '\u0000' else source[current + 1]
    private fun isAtEnd(): Boolean = current >= source.length
    private fun advance(): Char { current++; col++; return source[current - 1] }
}

File: src/lexer/lexer.rs
// src/lexer/lexer.rs
use crate::lexer::token::{Token, TokenKind};

pub struct Lexer {
    src: Vec<char>,
    pos: usize,
    line: usize,
    column: usize,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        Lexer {
            src: input.chars().collect(),
            pos: 0,
            line: 1,
            column: 1,
        }
    }

    pub fn tokenize(mut self) -> Vec<Token> {
        let mut tokens = Vec::new();
        loop {
            let tok = self.next_token();
            let is_eof = matches!(tok.kind, TokenKind::Eof);
            tokens.push(tok);
            if is_eof { break; }
        }
        tokens
    }

    fn next_char(&mut self) -> Option<char> {
        if let Some(&c) = self.src.get(self.pos) {
            self.pos += 1;
            if c == '\n' {
                self.line += 1;
                self.column = 1;
            } else {
                self.column += 1;
            }
            Some(c)
        } else {
            None
        }
    }

    fn peek_char(&self) -> Option<char> {
        self.src.get(self.pos).copied()
    }

    fn skip_whitespace(&mut self) {
        while let Some(c) = self.peek_char() {
            if c.is_whitespace() {
                self.next_char();
            } else {
                break;
            }
        }
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();
        let start_line = self.line;
        let start_col = self.column;
        let c = match self.next_char() {
            Some(ch) => ch,
            None => return Token::new(TokenKind::Eof, start_line, start_col),
        };

        // Range operator `..`
        if c == '.' && self.peek_char() == Some('.') {
            self.next_char();
            return Token::new(TokenKind::Range, start_line, start_col);
        }
        match c {
            '[' => Token::new(TokenKind::LBracket, start_line, start_col),
            ']' => Token::new(TokenKind::RBracket, start_line, start_col),
            '+' => Token::new(TokenKind::Plus, start_line, start_col),
            '-' => Token::new(TokenKind::Minus, start_line, start_col),
            '*' => Token::new(TokenKind::Star, start_line, start_col),
            '/' => Token::new(TokenKind::Slash, start_line, start_col),
            '%' => Token::new(TokenKind::Percent, start_line, start_col),
            '(' => Token::new(TokenKind::LParen, start_line, start_col),
            ')' => Token::new(TokenKind::RParen, start_line, start_col),
            '{' => Token::new(TokenKind::LBrace, start_line, start_col),
            '}' => Token::new(TokenKind::RBrace, start_line, start_col),
            ',' => Token::new(TokenKind::Comma, start_line, start_col),
            ':' => Token::new(TokenKind::Colon, start_line, start_col),
            '.' => Token::new(TokenKind::Dot, start_line, start_col),
            '?' => Token::new(TokenKind::Question, start_line, start_col),

            '=' => {
                if self.peek_char() == Some('=') {
                    self.next_char();
                    Token::new(TokenKind::EqEq, start_line, start_col)
                } else {
                    Token::new(TokenKind::Assign, start_line, start_col)
                }
            }
            '<' => {
                if self.peek_char() == Some('-') {
                    self.next_char();
                    Token::new(TokenKind::LeftArrow, start_line, start_col)
                } else if self.peek_char() == Some('=') {
                    self.next_char();
                    Token::new(TokenKind::Le, start_line, start_col)
                } else {
                    Token::new(TokenKind::Lt, start_line, start_col)
                }
            }
            '>' => {
                if self.peek_char() == Some('=') {
                    self.next_char();
                    Token::new(TokenKind::Ge, start_line, start_col)
                } else {
                    Token::new(TokenKind::Gt, start_line, start_col)
                }
            }
            '&' if self.peek_char() == Some('&') => {
                self.next_char();
                Token::new(TokenKind::AndAnd, start_line, start_col)
            }
            '|' if self.peek_char() == Some('|') => {
                self.next_char();
                Token::new(TokenKind::OrOr, start_line, start_col)
            }
            '!' if self.peek_char() == Some('=') => {
                self.next_char();
                Token::new(TokenKind::NotEq, start_line, start_col)
            }
            '!' => Token::new(TokenKind::Not, start_line, start_col),
            '#' => {
                // 跳过注释到行尾
                while let Some(nc) = self.peek_char() {
                    if nc == '\n' { break; }
                    self.next_char();
                }
                return self.next_token();
            }
            '"' => {
                let mut s = String::new();
                while let Some(nc) = self.next_char() {
                    if nc == '"' { break; }
                    if nc == '\\' {
                        if let Some(esc) = self.next_char() {
                            match esc {
                                'n' => s.push('\n'),
                                't' => s.push('\t'),
                                'r' => s.push('\r'),
                                '\\' => s.push('\\'),
                                '"' => s.push('"'),
                                other => { s.push('\\'); s.push(other); }
                            }
                            continue;
                        }
                        s.push('\\');
                        break;
                    }
                    s.push(nc);
                }
                Token::new(TokenKind::StringLiteral(s), start_line, start_col)
            }
            '\'' => {
                let ch = self.next_char().unwrap_or('\0');
                self.next_char();
                Token::new(TokenKind::CharLiteral(ch), start_line, start_col)
            }
            c if c.is_ascii_digit() => self.lex_number(c, start_line, start_col),
            c if c.is_alphabetic() || c == '_' => {
                let mut ident = c.to_string();
                while let Some(nc) = self.peek_char() {
                    if nc.is_alphanumeric() || nc == '_' {
                        ident.push(self.next_char().unwrap());
                    } else { break; }
                }
                let kind = match ident.as_str() {
                    "true" => TokenKind::BoolLiteral(true),
                    "false" => TokenKind::BoolLiteral(false),
                    // 关键字
                    kw @ "import" | kw @ "fun" | kw @ "async" | kw @ "await" |
                    kw @ "let" | kw @ "say" | kw @ "ask" | kw @ "as" |
                    kw @ "if" | kw @ "else" | kw @ "loop" | kw @ "forever" |
                    kw @ "return" | kw @ "break" | kw @ "continue" |
                    kw @ "in" | kw @ "bark" | kw @ "sniff" |
                    kw @ "snatch" | kw @ "lastly" | kw @ "nopaw" | 
                    kw @ "record" | kw @ "tail" => {
                        TokenKind::Keyword(kw.into())
                    }
                    // 类型
                    ty @ "Int" | ty @ "Long" | ty @ "Float" | ty @ "Double" |
                    ty @ "String" | ty @ "Char" | ty @ "Bool" | ty @ "Any" |
                    ty @ "Void" | ty @ "Array" => TokenKind::Type(ty.into()),
                    _ => TokenKind::Identifier(ident.clone()),
                };
                Token::new(kind, start_line, start_col)
            }
            _ => Token::new(TokenKind::Error(format!("Unexpected character: {}", c)), start_line, start_col),
        }
    }

    fn lex_number(&mut self, first: char, line: usize, col: usize) -> Token {
        let mut num = first.to_string();

        // 整数部分
        while let Some(c) = self.peek_char() {
            if c.is_ascii_digit() {
                num.push(self.next_char().unwrap());
            } else {
                break;
            }
        }

        // 如果是小数点，且不是范围操作符 ".."
        if self.peek_char() == Some('.') && self.src.get(self.pos + 1) != Some(&'.') {
            // 吃掉 '.'
            num.push(self.next_char().unwrap());
            // 小数部分
            while let Some(c2) = self.peek_char() {
                if c2.is_ascii_digit() {
                    num.push(self.next_char().unwrap());
                } else {
                    break;
                }
            }

            // 看看有没有后缀 f/F 或 d/D
            if let Some(suf) = self.peek_char() {
                match suf {
                    'f' | 'F' => {
                        // FloatLiteral(f32)
                        self.next_char();
                        match num.parse::<f32>() {
                            Ok(f) => return Token::new(TokenKind::FloatLiteral(f), line, col),
                            Err(_) => {
                                return Token::new(
                                    TokenKind::Error("Invalid float32 literal".into()),
                                    line,
                                    col,
                                )
                            }
                        }
                    }
                    'd' | 'D' => {
                        // DoubleLiteral(f64)
                        self.next_char();
                        match num.parse::<f64>() {
                            Ok(d) => return Token::new(TokenKind::DoubleLiteral(d), line, col),
                            Err(_) => {
                                return Token::new(
                                    TokenKind::Error("Invalid float64 literal".into()),
                                    line,
                                    col,
                                )
                            }
                        }
                    }
                    _ => {}
                }
            }

            // 无后缀，默认 DoubleLiteral
            match num.parse::<f64>() {
                Ok(d) => return Token::new(TokenKind::DoubleLiteral(d), line, col),
                Err(_) => {
                    return Token::new(
                        TokenKind::Error("Invalid float literal".into()),
                        line,
                        col,
                    )
                }
            }
        }

        // 长整型后缀 L 或 l
        if let Some(c) = self.peek_char() {
            if c == 'L' || c == 'l' {
                self.next_char();
                return match num.parse::<i64>() {
                    Ok(n) => Token::new(TokenKind::LongLiteral(n), line, col),
                    Err(_) => {
                        Token::new(TokenKind::Error("Invalid long literal".into()), line, col)
                    }
                };
            }
        }

        // 默认 int
        match num.parse::<i32>() {
            Ok(n) => Token::new(TokenKind::IntLiteral(n), line, col),
            Err(_) => Token::new(TokenKind::Error("Invalid int literal".into()), line, col),
        }
    }
}
